Source/santa-driver/SantaPrefixTree.cc

```diff
@@ -0,0 +1,147 @@
+
+bool SantaPrefixTree::init() {
+  if (!super::init()) return false;
+  root_ = new SantaPrefixNode;
+  root_->init();
+
+  spt_lock_grp_attr_ = lck_grp_attr_alloc_init();
+  spt_lock_grp_ = lck_grp_alloc_init("santa-prefix-tree-lock", spt_lock_grp_attr_);
+  spt_lock_attr_ = lck_attr_alloc_init();
+  spt_lock_ = lck_rw_alloc_init(spt_lock_grp_, spt_lock_attr_);
+
+  return true;
+}
+
+void SantaPrefixTree::AddPrefix(const char *prefix) {
+  LOGD("Trying to add prefix: %s", prefix);
+
+  lck_rw_lock_exclusive(spt_lock_);
+
+  SantaPrefixNode *node = root_;
+  size_t len = strlen(prefix);
+  for (int i = 0; i < len; ++i) {
+    // If there is a node in the path that is considered a prefix, stop adding.
+    // For our purposes we only care about the shortest path that matches.
+    if (node->isPrefix) break;
+
+    // Create a single char string.
+    const char value[2] = {prefix[i], '\0'};
+
+    // Create a new node if needed
+    if (!node->children->getObject(value)) {
+      SantaPrefixNode *newNode = new SantaPrefixNode;
+      newNode->init();
+      node->children->setObject(value, newNode);
+    }
+
+    // Get ready for the next iteration.
+    node = OSDynamicCast(SantaPrefixNode, node->children->getObject(value));
+
+    // If this is the end, mark the node as a prefix.
+    if (i + 1 == len) {
+      LOGD("Added prefix: %s", prefix);
+      node->isPrefix = true;
+      break;  // Unnecessary, but clear.
+    }
+  }
+
+  lck_rw_unlock_exclusive(spt_lock_);
+}
+
+bool SantaPrefixTree::HasPrefix(const char *string) {
+  lck_rw_lock_shared(spt_lock_);
+
+  auto found = false;
+
+  SantaPrefixNode *node = root_;
+  size_t len = strlen(string);
+  for (int i = 0; i < len; ++i) {
+    // Create a single char string.
+    const char value[2] = {string[i], '\0'};
+
+    // Get the next char string.
+    node = OSDynamicCast(SantaPrefixNode, node->children->getObject(value));
+
+    // If it doesn't exist in the tree, no match.
+    if (!node) break;
+
+    // If it does exist, is it a prefix?
+    if (node->isPrefix) {
+      found = true;
+      break;
+    }
+  }
+
+  lck_rw_unlock_shared(spt_lock_);
+
+  return found;
+}
+
+void SantaPrefixTree::free() {
+  if (spt_lock_) {
+    lck_rw_free(spt_lock_, spt_lock_grp_);
+    spt_lock_ = nullptr;
+  }
+
+  if (spt_lock_attr_) {
+    lck_attr_free(spt_lock_attr_);
+    spt_lock_attr_ = nullptr;
+  }
+
+  if (spt_lock_grp_) {
+    lck_grp_free(spt_lock_grp_);
+    spt_lock_grp_ = nullptr;
+  }
+
+  if (spt_lock_grp_attr_) {
+    lck_grp_attr_free(spt_lock_grp_attr_);
+    spt_lock_grp_attr_ = nullptr;
+  }
+
+  root_->release();
+  super::free();
+}
+
+OSDefineMetaClassAndStructors(SantaPrefixNode, OSObject);
+
+bool SantaPrefixNode::init() {
+  if (!super::init()) return false;
+  children = OSDictionary::withCapacity(1);
+  isPrefix = false;
+  return true;
+}
+
+void SantaPrefixNode::free() {
+  children->OSCollection::iterateObjects(^bool(OSObject *key) {
```

Comment:
This shouldn't, I believe, be necessary. All the OSCollection classes retain their keys and values when added and release when removed - as long as the free method calls release on the collection all of the objects in that collection will be released, cascading down.

Reply:
Without it the nodes are leaked :/
```
⇒  sudo kextunload -b com.google.santa-driver
(kernel) Can't unload kext com.google.santa-driver; classes have instances:
(kernel)     Kext com.google.santa-driver class SantaPrefixNode has 29 instances.
Failed to unload com.google.santa-driver - (libkern/kext) kext is in use or retained (cannot unload).
```
refcount = 0
`SantaPrefixNode *newNode = new SantaPrefixNode;`
refcount = 1
`node->children->setObject(value, newNode);`
refcount = 2

To release them:

`children->getObject((OSString *)key)->release();`
refcount = 1
`children->flushCollection();`
refcount = 0

The docs for `flushCollection` say:
>Removes and releases all keys and objects within the dictionary.
I take this as, "call release()" on all keys and objects.

Anyways, removing this code in favor of the char lookup array.

Reply:
That's because you're not releasing them after adding to the dictionary.

```
SantaPrefixNode *newNode = new SantaPrefixNode;  // newNode refcount = 1
children->setObject(value, newNode);  // newNode refcount = 2
newNode->release()  // newNode refcount = 1

...

children->release();  // newNode refcount = 0
```

Reply:
Ah, thats better.